// angular-oauth2-oidc/esm2015/events.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @abstract
 */
export class OAuthEvent {
    /**
     * @param {?} type
     */
    constructor(type) {
        this.type = type;
    }
}
if (false) {
    /** @type {?} */
    OAuthEvent.prototype.type;
}
export class OAuthSuccessEvent extends OAuthEvent {
    /**
     * @param {?} type
     * @param {?=} info
     */
    constructor(type, info = null) {
        super(type);
        this.info = info;
    }
}
if (false) {
    /** @type {?} */
    OAuthSuccessEvent.prototype.info;
}
export class OAuthInfoEvent extends OAuthEvent {
    /**
     * @param {?} type
     * @param {?=} info
     */
    constructor(type, info = null) {
        super(type);
        this.info = info;
    }
}
if (false) {
    /** @type {?} */
    OAuthInfoEvent.prototype.info;
}
export class OAuthErrorEvent extends OAuthEvent {
    /**
     * @param {?} type
     * @param {?} reason
     * @param {?=} params
     */
    constructor(type, reason, params = null) {
        super(type);
        this.reason = reason;
        this.params = params;
    }
}
if (false) {
    /** @type {?} */
    OAuthErrorEvent.prototype.reason;
    /** @type {?} */
    OAuthErrorEvent.prototype.params;
}
